README.md
# compliancelib-python
Machine readable cybersecurity compliance standards library for Python, starting with FISMA and NIST Risk Management Framework

Source code: https://github.com/govready/compliancelib-python

For more history, see earlier prototype 800-53-Control-Server (https://github.com/govready/800-53-server).

## Goal
Create a python class that generates FISMA 800-53 security control information that can be used by other software, including:

**Essential**
- [x] control identifier, title, description, supplemental guidence etc.
- [x] list of control enhancements identifiers
- [x] control responsibility (e.g., organization or information system)
- [x] related controls

**Better**
- [ ] control assignment parameters
- [x] output control information in various formats (e.g., JSON, YAML, OpenControl)
- [x] dependent controls (precursors) and recursive dependency trees
- [x] visual representation of control dependencies

**Best**
- [x] reading OpenControl files
- [x] load local repo standards, certifications, components
- [x] load local repo systems
- [x] resolving and reading OpenControl depencies
- [x] load remote repo standards, certifications, components, systems
- [x] system compliance status (e.g., control implementation narratives)
- [x] generating snippets for System Security Plan and other assessment artifacts
- [ ] mapping of control to (likely) generic organization role (e.g., developer, project manager, sys admin)

## Warnings
- This is pre-production and under active development.
- Reading of dependencies functions only one level deep (e.g., read indicated dependency files but no recursion)

## License
GPL 3.0

This was a lot of hardwork and has a number of firsts. Please support it or do not use it.

## Requirements
- Python 2.7 or 3.4+
- PyYAML
- graphviz
- defusedxml

## Installation
Compliancelib can be installed with Python pip

```
pip install compliancelib
```
To install Python pip:

- https://pip.pypa.io/en/stable/installing/
- http://sharadchhetri.com/2014/05/30/install-pip-centos-rhel-ubuntu-debian/

Compliancelib can beinstalled with Python Easy Install

```
    easy_install compliancelib
```

## Running tests

With python 2.7 (on a Mac):
(Note: you may need to include 'sudo' on a Mac, but that could also just be me)

```
sudo python setup.py test
```

With python 3.4+ (on a Mac):
(Note: you may need to include 'sudo' on a Mac, but that could also just be me)

```
sudo python3 setup.py test
```

## Usage

**Basic Usage**

Using ComplianceLib is straightforward. Almost everything centers around the control identifier.

Instantiate a security control object with ComplianceLib in a Python shell:

```python
>>> import compliancelib
>>> c = compliancelib.NIST800_53("AT-3")
>>> print("The title of {} is {}".format(c.id, c.title))
The title of AT-3 is ROLE-BASED SECURITY TRAINING
```

List various details of a control in a Python shell:

```python
>>> c.id
'AT-3'
>>> c.number
'AT-3'
>>> c.title
'ROLE-BASED SECURITY TRAINING'
>>> c.family
'AWARENESS AND TRAINING'
>>> c.description
'The organization provides role-based security training to personnel with assigned security roles and 
responsibilities:\na. Before authorizing access to the information system or performing assigned duties;\nb.
When required by information system changes; and\nc. [Assignment: organization-defined frequency] thereafter.'
>>> c.responsible
'organization'
>>> c.supplemental_guidance
'Organizations determine the appropriate content of security training based on the assigned roles and 
responsibilities of individuals and the specific security requirements of organizations and the information 
systems to which personnel have authorized access. In addition,  organizations provide enterprise architects,
...
>>> c.control_enhancements_textblock
'\nAT-3 (1)\nENVIRONMENTAL CONTROLS\nThe organization provides [Assignment: organization-defined personnel or roles] with initial and [Assignment: organization-defined frequency] training in the employment and operation of environmental controls.\nEnvironmental controls include, for example, fire suppression and detection devices/systems, sprinkler systems, handheld fire extinguishers, fixed fire hoses, smoke detectors,
...
>>> c.related_controls
['AT-2', 'AT-4', 'PL-4', 'PS-7' , 'SA-3' ,'SA-12', 'SA-16']
```

**Formatting**

ComplianceLib will also generate control information in various formats:

```python
>>> print(c.format('json'))
{"description": "The organization provides role-based security training to personnel with assigned security roles and 
responsibilities:\na. Before authorizing access to the information system or performing assigned duties;\nb. When required by 
information system changes; and\nc. [Assignment: organization-defined frequency] thereafter.", "title": "ROLE-BASED SECURITY 
...
responsibilities:", "description_sections": ["a. Before authorizing access to the information system or performing assigned 
duties;", "b. When required by information system changes; and", "c. [Assignment: organization-defined frequency] thereafter."]}

>>> print(c.format('yaml'))
description: 'The organization provides role-based security training to personnel
    with assigned security roles and responsibilities:

    a. Before authorizing access to the information system or performing assigned
    duties;

    b. When required by information system changes; and
...
title: ROLE-BASED SECURITY TRAINING

>>> print(c.format('control-masonry'))
description: The organization provides role-based security training to personnel with
    assigned security roles and responsibilities&colon; a. Before authorizing access
    to the information system or performing assigned duties; b. When required by information
    system changes; and c. [Assignment&colon; organization-defined frequency] thereafter.
description_intro: The organization provides role-based security training to personnel
    with assigned security roles and responsibilities&colon;
description_sections:
- a. Before authorizing access to the information system or performing assigned duties;
- b. When required by information system changes; and
- c. [Assignment&colon; organization-defined frequency] thereafter.
id: AT-3
name: ROLE-BASED SECURITY TRAINING
```

Example code for generating list of controls in `YAML` format:

```python
controllist = ["AT-3", "AU-1", "IR-2"]
d = dict()
for id in controllist:
    c = compliancelib.NIST800_53(id)
    d[id] = yaml.load(c.format('yaml'))

print(yaml.safe_dump(d, default_flow_style=False, encoding='utf-8', allow_unicode=True,
    explicit_start=True, explicit_end=True))
```

Example code for generating list of controls in `control-masonry` format:

```python
controllist = ["AT-3", "AU-1", "IR-2"]
d = dict()
for id in controllist:
    c = compliancelib.NIST800_53(id)
    d[id] = yaml.load(c.format('control-masonry'))

print(yaml.safe_dump(d, default_flow_style=False, encoding='utf-8', allow_unicode=True,
    explicit_start=True, explicit_end=True))

```

**Advanced Usage - Dependencies**

ComplianceLib's `NIST800_53Viz` class creates a graph of all precursor controls for a given control. ComplianceLib is the first time these precursor controls have been made available as code.

The `NIST800_53Viz` class will also generate a graphviz file visualizing nodes and edges of the dependency graph for a security control.

The list of precursor controls are extracted from [NIST SP 800-53 R1 Assessment Cases](compliancelib/data/800-53A-R1_Assessment-Cases_All-18-Families_ipd). We extracted the precursor-controls from NIST Assessment Guide documents into simplified data structure listing precursor, concurrent, and successor controls by family. View these files in this repo's [compliancelib/data/dependencies](compliancelib/data/dependencies) subdirectory.

To see control dependencies, simply do in python shell::

```python
>>> import compliancelib
>>> cv = compliancelib.NIST800_53Viz("AU-3")
>>> cv.precursor_controls
['AU-3', 'AU-2', 'RA-3', 'PM-9']
```

Creating the graphviz file is currently left as a reader exercise until future documentation completed.

## Compliance as Code

**Expressing security controls as code is useful.**
**Expressing system compliance as code is a game-changer.**

[OpenControl](http://open-control.org) is an emerging "Compliance as Code" community developing open-source, re-usable, shared compliance-by-component information and support tools. The goal is to allow developers to represent compliance as code of their component libraries and assembled systems in maintained repositories.

ComplianceLib's `OpenControlClass` and `SystemCompliance` classes work together to read OpenControl data files and represent an Information System's compliance state as a Python object that can be queried.

The `OpenControlClass` and `SystemCompliance` classes are under heavy development in ComplianceLib versions 0.8.0 through versions 0.15.0 with class attributes and methods subject to significant change.

Below is an example of using ComplianceLib to load and query compliance posture of the OpenControl [Freedonia-Compliance](https://github.com/opencontrol/freedonia-compliance) tutorial.

```python
>>> import compliancelib
>>> sp = compliancelib.SystemCompliance()
>>> sp.load_system_from_opencontrol_repo('https://github.com/opencontrol/freedonia-compliance')
[LOG compliancelib]; INFO; 2016-10-19 06:05:57,807; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-compliance/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,503; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-compliance/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,503; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-compliance/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,503; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-compliance/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,503; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-compliance/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,503; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-compliance/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,503; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-frist/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,641; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-frist/master/opencontrol.yaml
[LOG compliancelib]; INFO; 2016-10-19 06:05:58,842; opencontrolfiles; ocfileurl: https://raw.githubusercontent.com/opencontrol/freedonia-aws-compliance/master/opencontrol.yaml
True

>>> sp.system['name'] = "My Awesome System"
>>> sp.system['name']
'My Awesome System'

>>> sp.components()
['Audit Policy', 'AWS Core', 'AWS Implementation']
>>> sp.standards()
['FRIST-800-53']
>>> sp.certifications()
['FredRAMP-low']
>>> sp.systems()
['freedonia-aws']


>>> sp.control('AU-1').title
'AUDIT AND ACCOUNTABILITY POLICY AND PROCEDURES'
>>> sp.control('AU-1').description
'The organization:\na. Develops, documents, and disseminates to [Assignment: organization-defined personnel or roles]:\na.1. An audit and accountability policy that addresses purpose, scope, roles, responsibilities, management commitment, coordination among organizational entities, and compliance; and\na.2. Procedures to facilitate the implementation of the audit and accountability policy and associated audit and accountability controls; and\nb. Reviews and updates the current:\nb.1. Audit and accountability policy [Assignment: organization-defined frequency]; and\nb.2. Audit and accountability procedures [Assignment: organization-defined frequency].'
>>> sp.control('AU-1').priority
'P1'
>>> sp.control('AU-1').implementation_status
['implemented']
>>> sp.control('AU-1').implementation_status_details
{'Audit Policy': 'implemented'}
>>> sp.control('AU-1').components
['Audit Policy']
>>> sp.control('AU-1').components_dict
{'Audit Policy': [{'narrative': [{'text': 'This text describes how our organization is meeting the requirements for the\nAudit policy, and also references a more complete description at ./AU_policy/README.md\n\nSince the AU-1 `control` is to document and disseminate a policy on Audit and Accountability, then\nthis narrative suffices to provide that control. A verification step could be something\nthat checks that the referenced policy is no more than 365 days old.\n'}], 'control_key': 'AU-1', 'covered_by': [], 'standard_key': 'FRIST-800-53', 'implementation_status': 'implemented'}]}
```

To import a local repo:

```
sp.load_system_from_opencontrol_repo('file:///fullpath/to/localfile/freedonia-compliance')
```

Adjust commandline verbosity by set log level of OpenControlFile class to one of CRITICAL, ERROR, WARNING, INFO, DEBUG:

```
>>> ocf = compliancelib.OpenControlFiles()
>>> ocf.logger.setLevel("DEBUG")
>>> ocf.logger.setLevel("CRTICAL")
```

Looking at the `sp.control` object dictonary provides a glimpse of the roadmap:

```python
>>> sp.control('AU-1').__dict__.keys()
dict_keys(['responsible', 'implementation_status_details', 'implementation_status', 'title', 'related_controls', 'id', 'control_enhancements', 'description_sections', 'components_dict', 'json_dict', 'assignments', 'implementation_narrative', 'family', 'description', 'control_enhancements_textblock', 'supplemental_guidance', 'components', 'description_intro', 'sg', 'priority', 'validation', 'number', 'roles'])
```

The roadmap includes emitting text snippets for System Security Plans:

```python
>>> sp.control_ssp_text('AU-1')
AU-1 - AUDIT AND ACCOUNTABILITY POLICY AND PROCEDURES
The organization:
a. Develops, documents, and disseminates to [Assignment: organization-defined personnel or roles]:
a.1. An audit and accountability policy that addresses purpose, scope, roles, responsibilities, management commitment, coordination among organizational entities, and compliance; and
a.2. Procedures to facilitate the implementation of the audit and accountability policy and associated audit and accountability controls; and
b. Reviews and updates the current:
b.1. Audit and accountability policy [Assignment: organization-defined frequency]; and
b.2. Audit and accountability procedures [Assignment: organization-defined frequency].


responsible: organization
roles: {}
implementation status: ['implemented']


via Audit Policy
This text describes how our organization is meeting the requirements for the
Audit policy, and also references a more complete description at ./AU_policy/README.md

Since the AU-1 `control` is to document and disseminate a policy on Audit and Accountability, then
this narrative suffices to provide that control. A verification step could be something
that checks that the referenced policy is no more than 365 days old.
```

CHANGELOG.md
# Change Log

## compliancelib v1.1.2
- fix OpenControlFiles crash when OpenControl repo contains no dependencies

## compliancelib v1.1.1
- Improve OpenControl narrative parsing to handle narratives that are strings and lists

## compliancelib v1.1.0
- OpenControlFiles class generates list of dependencies by item type
- SystemCompliance class consumes content listed in dependencies (but only one level deep, e.g., will retrieve item types that are locally listed in the dependent repository, but will not recursive deeper)
- Technical debt added of possible duplicate keys read from remote repos are not tested again existing keys (e.g., a local component file and dependent component file referring to same ID for component could collide)

## complianceLib v1.0.0
- Refactor OpenControlFiles class to read components, standards, certifications from shared methods
- SystemCompliance reading standards and certificaftions
- NOTE: SystemCompliance not yet consuming content from remote dependency repos into

## compliancelib v0.13.6
- Add python logging to OpenControlFiles class

## compliancelib v0.13.4
- Indicate local repos supported in error message

## compliancelib v0.13.3
- No changes, bumping version number

## compliancelib v0.13.2
- Resolve opencontrol components on localfile system, fixing component references

## compliancelib v0.13.1
- Resolve opencontrol components on localfile system

## compliancelib v0.13.0
- Resolve opencontrol repos on localfile system

## compliancelib v0.12.2
- Correcting link in README.rst

## compliancelib v0.12.1
- Add GPL 3.0 license

## compliancelib v0.12.0
- Update README
- Represent related controls as array instead of string
- Improve tests for related control cases

## compliancelib v0.11.0
- Represent a control's enhancements as an array of control enhancement ids
- Move textblock of control enhancements into attribute control_enhancement_textblock
- Rename internal method for list of all control enhancements in 800-53

## compliancelib v0.10.0
- Changed name of `list_component*` methods

## compliancelib v0.9.0
- Add OpenControlFiles class to read open controls
- Resolve and ingest opencontrol.yaml file from GitHub repos
- Resolve and ingest OpenControl component yaml files (no dependencies)
- Update SystemCompliance class to load components via OpenControl YAML reference
- Update SystemCompliance to dump plain text write-up of NIST800-53 System Security Plan control implementation write-up
- NOTE: ingesting OpenControlFiles still incomplete

## compliancelib v0.8.0
- Add SystemCompliance class to represent System Compliance
- System Compliance class consumes OpenControl formatted component files for control implementation

## compliancelib v0.7.4
- handle control enhancement that missing data elements

## compliancelib v0.7.3
- replace xml.etree.ElementTree with defusedxml.ElementTree
- remove unused imports

## compliancelib v0.7.2
- Improve json nosetest
- rename test files to nist800_53 convention
- compliancelib.NIST800_53Viz v0.3.1
	- Use 2to3 to upgrade nist800_53.py to Python3

## compliancelib v0.7.0
ALERT: Renaming of SecControl to NIST800_53
- Rename compliancelib.SecControl to compliancelib.NIST800_53
- Rename compliancelib.SecControlViz to compliancelib.NIST800_53Viz

## compliancelib v0.6.1
- Correct changelog of compliancelib v0.6.0

## compliancelib v0.6.0
- compliancelib.SecControl v0.7.0
	- Add `format` method to produce `JSON`, `YAML`, `Control-Masonry` format
- compliancelib.SecControlViz v0.2.0
	- no changes

## compliancelib v0.5.0
- compliancelib.SecControl v0.6.0
	- no changes
- compliancelib.SecControlViz v0.2.0
	- Add attribute `precursor_controls`

## compliancelib v0.4.2
- compliancelib.SecControl v0.6.0
	- Remove `xsltproc` dependency, parse XML with Python `xml.etree.ElementTree`
	- Add CHANGELOG.md file
	- README improvements


compliancelib/systemcompliance.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""Class for OpenControl

Instantiate and then...

Load an OpenControl component YAML file (e.g., https://raw.githubusercontent.com/pburkholder/freedonia-compliance/master/AU_policy/component.yaml).

Methods provide information about the Security Control.

Visit [tbd] for the latest version.

LICENSE

ComplianceLib SystemCompliance is a class for representing compliance as code for an information system.
Copyright (C) 2016  GovReady PBC.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


Example python CLI
------------------

import sys, yaml, pprint

import compliancelib
# instantiate an SystemCompliance object
sp = compliancelib.SystemCompliance()

# change name
sp.system['name'] = "GovReady WordPress Dashboard"

# Idempotent adding components from URLs
sp.components()
f1 = 'https://raw.githubusercontent.com/pburkholder/freedonia-compliance/master/AU_policy/component.yaml'
f2 = 'https://raw.githubusercontent.com/opencontrol/cf-compliance/master/UAA/component.yaml'
sp.add_component_from_url(f1)
sp.add_component_from_url(f2)
sp.add_component_from_url(f2)
sp.components()
# Alternatively
sp.add_system_dict_from_url('components', f1)
sp.add_system_dict_from_url('components', f2)
sp.components()
# result
# ['Audit Policy', 'User Account and Authentication (UAA) Server']

# display summary info
sp.summary()
# result
# {'stanards': [], 'certifications': [], 'name': 'GovReady WordPress Dashboard', 'components': ['Audit Policy', 'User Account and Authentication (UAA) Server']}
pprint.pprint(sp.summary())
# result
# {'certifications': [],
# 'components': ['Audit Policy',
#                'User Account and Authentication (UAA) Server'],
# 'name': 'GovReady WordPress Dashboard',
# 'stanards': []}

# query cntrol
ck = "AC-4"
ci = sp.control(ck)
ci.components
ci.components_dict
ci.implementation_status
ci.assignments
ci.implementation_narrative
print(ci.implementation_narrative)

# Test loading GovCloud OpenControl component yaml directly
component_list = ['AC_Policy','AT_Policy','AU_Policy','CA_Policy','CICloudGov','CM_Policy','CP_Policy','CloudCheckr','ELKStack','IA_Policy','IR_Policy','JumpBox','MA_Policy','MP_Policy','PE_Policy','PL_Policy','PS_Policy','RA_Policy','SA_Policy','SC_Policy','SI_Policy','SecureProxy']
urls = ["https://raw.githubusercontent.com/18F/cg-compliance/master/%s/component.yaml" % comp for comp in component_list]
for compurl in urls:
  sp.add_component_from_url(compurl)

# nice controls to test: AU-1, AU-5, SC-10

# Print out control details
ck = "AC-2 (1)"
ci = sp.control(ck)
ci.components
ci.components_dict
print(ci.implementation_narrative)


# very short, load from opencontrol
import compliancelib
sp = compliancelib.SystemCompliance()
sp.load_system_from_opencontrol_repo('https://github.com/18F/cg-compliance')
sp.control('AC-4').title
print(sp.control('AC-4').description)
print(sp.control('AC-4').implementation_narrative)
print(sp.control('AC-4').implementation_status)
sp.control_ssp_text('AC-4')

# list controls from each component
for component in sp.components():
  print(component)
  [ck['control_key'] for ck in sp.system['components'][component]['satisfies']]

# freedonia-compliance
import compliancelib
sp = compliancelib.SystemCompliance()
sp.load_system_from_opencontrol_repo('https://github.com/opencontrol/freedonia-compliance')
sp.control('AU-1').title
sp.control_ssp_text('AU-1')
print(sp.control('AU-1').description)
print(sp.control('AU-1').implementation_narrative)

for component in sp.components():
  print(component)
  [ck['control_key'] for ck in sp.system['components'][component]['satisfies']]
"""

__author__ = "Greg Elin (gregelin@govready.com)"
__version__ = "$Revision: 1.1.1 $"
__date__ = "$Date: 2016/12/18 20:13:00 $"
__copyright__ = "Copyright (c) 2016 GovReady PBC"
__license__ = "GNU General Public License v3 (GPLv3)"

import os
import json
import yaml
import re
import sys
from .nist800_53 import NIST800_53
from .opencontrolfiles import OpenControlFiles

if sys.version_info >= (3, 0):
    from urllib.parse import urlparse
    from urllib.request import urlopen
if sys.version_info < (3, 0) and sys.version_info >= (2, 5):
    from urlparse import urlparse
    from urllib2 import urlopen

class SystemCompliance():
    "initialize SystemCompliance security controls implementation"
    def __init__(self):
        self.ocfiles = {}
        # define the dictionaries we will support so we avoid unexpected data
        self.supported_dictionaries = ['components', 'standards', 'certifications', 'assignments', 'roles', 'systems']
        # stub out a system
        self._stub_system()
        pass

    def _stub_system(self):
      # Load a stub file of the sytem
      self.system = {}
      self.system['name'] = ""
      self.system['components'] = {}
      self.system['certifications'] = {}
      self.system['standards'] = {}
      self.system['systems'] = {}
      self.system['assignments'] = {}
      self.system['roles'] = {}

    def system_component_add(self, component_name, component_dict):
      "add a component as a dictionary to the system with the component name as key"
      self.system['components'][component_name] = component_dict

    def add_component_from_url(self, oc_componentyaml_url):
      "add a component as a dictionary to the system from an OpenControl YAML file at a URL"
      # go load opencontrol file
      print("oc_componentyaml_url: ", oc_componentyaml_url)
      try:
        my_dict = yaml.safe_load(urlopen(oc_componentyaml_url))
        # todo - add checks to make sure it is a proper opencontrol file
      except:
        print("Unexpected error loading YAML file:", sys.exc_info()[0])
        my_dict = None
        raise
      if (my_dict):
        # TODO: Consider if method `system_component_add` needed
        self.system_component_add(my_dict['name'], my_dict)

    def components(self):
      "list the components composing the system as array"
      return list(self.system['components'])

    # generic method for loading dictionaries
    def add_system_dict(self, my_dict_type, my_dict_name, my_dict):
      "load a dictionary into system object"
      if my_dict_type in self.supported_dictionaries:
        # to do - validate dictionary
        if (my_dict and my_dict_name):
          self.system[my_dict_type][my_dict_name] = my_dict
      else:
        # pass or indicate error
        raise Exception('Attempt to load unsupported dictionary type %s' % my_dict_type)

    def add_system_dict_from_url(self, dict_type, url):
      "load a dictionary into system object from a URL"
      if dict_type in self.supported_dictionaries:
        try:
          my_dict = yaml.safe_load(urlopen(url))
          # todo - validate proper opencontrol file
        except:
          print("Unexpected error loading YAML file:", sys.exc_info()[0])
          my_dict = None
          raise
        if (my_dict):
          self.system[dict_type][my_dict['name']] = my_dict
      else:
        raise Exception('Attempt to load unsupported dictionary type %s' % dict_type)

    def standards(self):
      "list the standards composing the system as array"
      return list(self.system['standards'])

    def certifications(self):
      "list the certifications composing the system as array"
      return list(self.system['certifications'])

    def systems(self):
      "list the systems composing the system as array"
      return list(self.system['systems'])

    def roles(self):
      "list the roles composing the system as array"
      return list(self.system['roles'])

    def summary(self):
      "dump high level system compliance profile abstract"
      scpa = {"name" :  self.system['name'],
        "components" : list(self.system['components']),
        "standards" : list(self.system['standards']),
        "certifications" : list(self.system['certifications'])}
      return scpa

    def control(self, cid, standard = 'NIST800_53'):
      "create a control object for system combining control info from standard and implementation details"
      # raise error if components are not loaded
      if len(self.components()) < 1:
        raise Exception ("No controls available. No components have been loaded.")

      # create control object from standard (NOTE: only support NIST 800-53 currently)
      if (standard == 'NIST800_53'):
        ci = NIST800_53(cid)
      else:
        raise Exception ("The standard %s is not currently supported." % standard)

      # find all components with content regarding control_key
      # add dictionary of components attribute to control
      ci.components_dict = {}
      for component in self.components():
        component_control_info = [ck for ck in self.system['components'][component]['satisfies'] if ck['control_key'] == cid]
        if (len(component_control_info)) > 0:
          ci.components_dict[component] = component_control_info

      # make it easy to look at list of components related to control by getting list of keys that are names of components
      ci.components = list(ci.components_dict)

      # determine roles
      ci.roles = {}

      # determine assignments
      ci.assignments = {}

      # determine narrative
      ci.implementation_narrative = ""
      ctl_str = ""
      for component in ci.components:
        # print component
        comp_contribution  = ci.components_dict[component][0]
        # Check if type(comp_contribution) is string or list
        if type(comp_contribution['narrative']) is str:
          # comp_contribution['narative'] is a string, like this:
          # Ex: {'control_key': 'AC-4', 'standard_key': 'NIST-800-53', 'covered_by': [], 'implementation_status': 'none', 'narrative': 'Cloud.Gov enforces security groups and other network traffic rules in a strict priority order. Cloud.Gov returns an allow...
          text = comp_contribution['narrative']
          ctl_str += "via %s\n%s\n" % (component, text)
        if type(comp_contribution['narrative']) is list:
          # comp_contribution['narative'] is a list, like this:
          # {'control_key': 'AC-4', 'standard_key': 'NIST-800-53', 'covered_by': [], 'implementation_status': 'none', 'narrative': [{'text': 'The information system enforces approved authorizations for controlling the flow of information within the system
          for text_item in comp_contribution['narrative']:
            key = ""
            text = ""
            if 'key' in text_item:
              key = "%s:\n" % text_item['key']
            elif 'text' in text_item:
              text = text_item['text']
            # Todo - deal with error case
            else:
              # we have a likely error in the data formating of the opencontrol file
              # Report an error
              text = "Possible error in %s's OpenControl file, no identified 'text' key in narrative block"
            ctl_str += "%s\nvia %s\n%s\n" % (key, component, text)
      ci.implementation_narrative = ctl_str

      # determine implementation status
      # We need a dictionary because mutiple components support a control
      ci.implementation_status_details = {}
      for component in ci.components:
        comp_contribution  = ci.components_dict[component][0]
        ci.implementation_status_details[component] = comp_contribution['implementation_status']
      # now that we have implementation status details for components, calculate actual status
          # From FedRAMP-SSP-Template-High-2016-06-20-v01-00.docx
          # Implementation Status (check all that apply):
          # ☐ Implemented
          # ☐ Partially implemented
          # ☐ Planned
          # ☐ Alternative implementation
          # ☐ Not applicable
      ci.implementation_status = [ci.implementation_status_details[component] for component in list(ci.implementation_status_details)]
      # determine validation
      ci.validation = {}

      # return the control implementation object
      return ci

    def control_ssp_text(self, cid):
      "print out text for a control listing in system security plan (assume NIST800-53)"
      ci = self.control(cid)
      print("%s - %s" % (ci.id, ci.title))
      print("%s" % (ci.description))
      print("\n")
      print("responsible: %s" % ci.responsible)
      print("roles: %s" % ci.roles)
      print("implementation status: %s\n" % ci.implementation_status)
      print("%s" % ci.implementation_narrative)

    def load_system_from_opencontrol_repo(self, repo_url, revision='master', verbose=''):
      "load system details and control implementation from a repo"
      # TODO Reset all values before loading new system
      # TODO handle not finding opencontrol.yaml file in repo
      ocf =  OpenControlFiles()

      for item_type in self.supported_dictionaries:
        for url in ocf.list_items_urls_in_repo(ocf.resolve_ocfile_url(repo_url, revision), item_type):
          if (verbose=='v'): print("Reading %s %s" % (item_type, url))
          self.add_system_dict_from_url(item_type, url)

      # load dependencies
      for item_type in self.supported_dictionaries:
        # get standards in dependencies
        for dependency_item in ocf.list_dependency_items_in_repo(ocf.resolve_ocfile_url(repo_url, revision), item_type):
          dependency_ocf =  OpenControlFiles()
          dependency_repo_url = dependency_item['url']
          dependency_revision = dependency_item['revision']
          if (item_type == 'systems'):
            url = dependency_ocf.resolve_ocfile_url(dependency_repo_url.strip("/"), dependency_revision)
            self.add_system_dict_from_url(item_type, url)
            # read remote item_type components
            for url in dependency_ocf.list_items_urls_in_repo(dependency_ocf.resolve_ocfile_url(dependency_repo_url.strip("/"), dependency_revision), "components"):
              # TODO test if we will end duplicating and overwriting existing key
              self.add_system_dict_from_url("components", url)
          else:
            # read remote item_type indicated
            for url in dependency_ocf.list_items_urls_in_repo(dependency_ocf.resolve_ocfile_url(dependency_repo_url.strip("/"), dependency_revision), item_type):
              # TODO test if we will end duplicating and overwriting existing key
              self.add_system_dict_from_url(item_type, url)

      return True

